### ISSUE TITLE
Set iteration can return booleans

### ISSUE THREAD

**Issue by asmeurer** [sympy/sympy#16362]
For example 

```py
>>> list(Intersection(FiniteSet(x), FiniteSet(y)))
[Contains(x, {y}), Contains(y, {x})]
>>> list(Intersection(FiniteSet(x), Reals))
[(x > -oo) & (x < oo)]
```

The problem here is that `Contains` and `And` are booleans, so they aren't actually elements of the set (the elements of the set are numbers, not booleans). 

See also https://github.com/sympy/sympy/pull/10249.

I would suggest raising a TypeError in this case. That would match the behavior of `in` (`__contains__`) and `bool` on relationals (like `bool(x < y)`).

I initially had a commit to do this at https://github.com/sympy/sympy/pull/16344/commits/ff65f86f9da9d848805d944c6d02f9d41c791ace, but it leads to some test failures in the integration tests, because `_eval_interval` iterates over the singularities, which are from solveset.

I think we need a way to iterate over elements that isn't strict (but doesn't return booleans, that is just mathematically wrong), and a way that is. I'm not sure which should be the default for `__iter__`. 

Also one comment on the above commit, it should be TypeError, not ValueError, to match `Set.__contains__` and `Relational.__bool__`.

**Comment by smichr** [comment]
These are now unevaluated:
```python
>>> Intersection(FiniteSet(x), Reals)
Intersection({x}, Reals)
>>> Intersection(FiniteSet(x), FiniteSet(y))
Intersection({x}, {y})
```

**Comment by oscarbenjamin** [comment]
Unless I'm misunderstanding this is still an issue on current master:
```julia
In [2]: list(Intersection(FiniteSet(x), FiniteSet(y)))                                                                            
Out[2]: [x ∈ {y}, y ∈ {x}]

In [3]: list(Intersection(FiniteSet(x), Reals))                                                                                   
Out[3]: [x > -∞ ∧ x < ∞]
```
The bug is presumably in `Intersection.__iter__`...

**Comment by asmeurer** [comment]
Yes, the commit I mentioned can be used as a starting point. But it caused some test failures which will have to be worked out. We need to figure out how we want to handle strict vs. non-strict set iteration first, though.

**Comment by sylee957** [comment]
I would reopen this issue because I think that there can be a better way to expand a set containment logic as relationals.

**Comment by oscarbenjamin** [comment]
@sylee957 I'm not sure what you mean. We have this:
```julia
In [5]: Intersection(FiniteSet(x), FiniteSet(y)).as_relational(z)                                                                              
Out[5]: z = x ∧ z = y

In [6]: Intersection(FiniteSet(x), Reals).as_relational(z)                                                                                     
Out[6]: z = x ∧ -∞ < z ∧ z < ∞
```
Do you think that something else should be possible?

AFAICT the issue as described above seems to be fixed.

**Comment by sylee957** [comment]
If we already have `Contains`, we should have rewriting methods like

![image](https://user-images.githubusercontent.com/34944973/65559183-988ba600-df74-11e9-908b-5d298e19127c.png)

and I think that `as_relational` is more naive idea than that.
